### DISCUSSION TITLE
Continuation of the discussion #15802

### DISCUSSION THREAD

**Discussion by PatLub85** [micropython/micropython#15869] [General]
Hi,
I have a small robot car project that works either autonomously or controlled with a Microdot application.
The switch between the two programs is done with a push button.
The model works but when I import Microdot I get the following message:
```
Traceback (most recent call last):
  File "<stdin>", line 8, in <module>
  File "prog_1_2.py", line 10, in <module>
MemoryError: memory allocation failed, allocating 2344 bytes
```
The line 10
```
from microdot import Microdot,Response,send_file
```
As stated on this page (https://docs.micropython.01studio.org/zh-cn/latest/reference/constrained.html), I wanted to add constants
The first one I tried was this:
```
# line 14
_PIN1 = const(Pin(32, Pin.OUT))
```
and I received this :
```
Traceback (most recent call last):
  File "<stdin>", line 14
SyntaxError: not a constant
```
A pin number can't be a constant?

I would like to freeze the code I have not managed to understand the process for an ESP32 under Windows.
Where can I find a tutorial that explains all the steps
I have seen examples with ESP_IDF tools, other examples with manifest files
I'm lost
Another solution would be to buy another board with more memory, but I would like to take advantage of this case to learn

**Comment by GitHubsSilverBullet** [comment]
> A pin number can't be a constant?

An integer number can.
A Pin Object can't.

**Reply by PatLub85** [reply]
ok

**Comment by scruss** [comment]
Also, that link you gave is to the documentation for the 01Studio MicroPython fork. It has some features that regular MicroPython doesn't have. Please refer to the correct documentation: [Overview — MicroPython v1.23.0 documentation](https://docs.micropython.org/en/v1.23.0/)

**Reply by PatLub85** [reply]
Ok thanks. It is true that we get a little lost with all these information sites.
Does anyone have a simple solution for the code freeze?

**Comment by sophiedegran** [comment]
If you want to freeze Microdot 
simply in manifest.py add:
package ('microdot', base_path='path_your_directory_microdot')
if you want to freeze a x.py prog
simply add:
freeze ('path_to_your_file', 'x.py')
Hope that help

**Reply by PatLub85** [reply]
Hi,
I'm sorry but I don't know where to start.
I know Python quite well, but much less about hardware.
I'm on Windows 11, Thonny or Vscode, ESP32S .
From there, what should I do to create the manifest and then compile the code. Until now I went to the Micropython site and flashed my card directly with Thonny.
You understand my level regarding hardware

**Reply by shariltumin** [reply]
The microdot http server is large, so you got a MemoryError.

Precompile the package files into mpy. Use mpy-cross.

You don't need a full HTTP server; just a TCP socket to receive commands for your robot car. If you listen to port 23, you can use any telnet client to send commands like "turn:left:90", "forward:10", and "stop".

Learning mpy-cross and mremote can make working with MicroPython easier. Remember that MicroPython runs on a resource-constrained MCU, so keep it simple.

**Reply by PatLub85** [reply]
I'm slowly moving forward with a lot of difficulty. Here's where I am:
I installed the mpy-cross plugin with Thonny
If I do
```
import mpy_cross
Traceback (most recent call last):
File "<stdin>", line 1, in <module>
ImportError: no module named 'mpy_cross'
```
I didn't get discouraged and found mpy-cross.exe in the \thonny\plugins\python310\site-packages directory
I did mpy-cross test_prog_1.py with the Python console and got a .mpy file
Hooray!
```
#test_prog_1.py
import time
def prog_1():
    while True:
        print('program 1')
        time.sleep(1)
        
if __name__ == "__main__":
    prog_1()
```
```
>>> import test_prog_1
It's ok
>>> prog_1()

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'prog_1' isn't defined
```
What did I miss?

**Reply by GitHubsSilverBullet** [reply]
>>> import test_prog_1
>>> prog_1()
> 
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> NameError: name 'prog_1' isn't defined
> ```
> 
> What did I miss?

```python
import test_prog_1
test_prog_1.prog_1()
```
or:
```python
from test_prog_1 import prog_1
prog_1()
```

**Reply by PatLub85** [reply]
It's ok, sorry for this gross mistake. As they say at home, I have my head down
Otherwise is my approach correct or is there a simpler approach?

**Reply by shariltumin** [reply]
Make sure that the file "test_prog_1.mpy" was loaded to your board. I don't know how to do this in Thonny.

Now if you have both "test_prog_1.py" and "test_prog_1.mpy" on your board, the ```import test_prog_1``` will read and use the "test_prog_1.py" and NOT the "test_prog_1.mpy". So if you want to use the compiled version i.e. "test_prog_1.mpy"  make sure you delete the source i.e. the "test_prog_1.py" from your board.

The test ```if __name__ == "__main__":``` is a carry-over from the CPython world that makes "test_prog_1.py" a standalone program. If you have a Linux or Windows port of MicroPython, you can do this:

```
micropython test_prog_1.py
program 1
program 1
program 1
^C
```

If you try to do this with "test_prog_1.mpy" you will get error:
```
micropython test_proc_1.mpy
Traceback (most recent call last):
  File "test_proc_1.mpy", line 1
SyntaxError: invalid syntax
```

I have never used the ```if __name__ == "__main__":``` switch in my MicroPython scripts.

I think this might be a great time to learn **mpremote**. If you're in a directory with "test_prog_1.mpy," you can just do:

```
mpremote mount . exec "import test_proc_1; test_proc_1.prog_1()"
```
or

```
mpremote mount . exec "from test_proc_1 import prog_1; prog_1()"
```

The only downside is that you have to work in a terminal window, not an IDE.

Why is this important? You don't have to load your script to the board during this process, which saves your board flash memory life. As you know, flash memory has a limited write cycle.

I see there's an "mpremote" plug-in for Thonny, but I haven't tried it myself, so I can't say for sure.

Anyway, I'm happy you were able to get your scripts to compile into mpy.

**Reply by PatLub85** [reply]
After transforming microdot.py into mpy, I have no more memory problems.
Now on your advice I will look at mpremote which seems very interesting
